pytential/symbolic/execution.py

```diff
@@ -544,23 +500,19 @@ def matvec(self, x):
         #    => output is a flat PyOpenCL array
         # * structured arrays (object arrays/DOFArrays)
         #    => output has same structure as input
-        if isinstance(x, np.ndarray) and x.dtype.char != "O":
-            x = self.array_context.from_numpy(x)
-            flat = True
-            host = True
-            assert x.shape == (self.total_dofs,)
+        if isinstance(x, DOFArray):
+            flat, host = False, False
+        elif isinstance(x, np.ndarray) and x.dtype.char == "O":
+            flat, host = False, False
         elif isinstance(x, cl.array.Array):
-            flat = True
-            host = False
+            flat, host = True, False
+            assert x.shape == (self.total_dofs,)
+        elif isinstance(x, np.ndarray) and x.dtype.char != "O":
+            x = self.array_context.from_numpy(x)
+            flat, host = True, True
             assert x.shape == (self.total_dofs,)
-        elif isinstance(x, np.ndarray) and x.dtype.char == "O":
-            flat = False
-            host = False
-        elif isinstance(x, DOFArray):
-            flat = False
-            host = False
         else:
-            raise ValueError("unsupported input type")
+            raise ValueError(f"unsupported input type: {type(x).__name__}")
```

Comment:
This is obviously unrelated. It just rearranges the order of the if statements so that the common `DOFArray` case is on top.